exporter/xray/segment.go

```diff
@@ -0,0 +1,319 @@
+	Annotations map[string]interface{} `json:"annotations,omitempty"`
+
+	// SubSegments contains the list of child segments
+	SubSegments []*segment `json:"subsegments,omitempty"`
+
+	// Service - optional service definition
+	Service *service `json:"service,omitempty"`
+}
+
+type service struct {
+	// Name of the service4
+	Name string `json:"name,omitempty"`
+
+	// Version - A string that identifies the version of your application that served the request.
+	Version string `json:"version,omitempty"`
+}
+
+type http struct {
+}
+
+// makeID - returns a 64-bit identifier for the segment, unique among segments in
+// the same trace, in 16 hexadecimal digits.
+func makeID(spanID trace.SpanID) string {
+	if spanID == zeroSpanID {
+		return ""
+	}
+	return hex.EncodeToString(spanID[0:8])
+}
+
+// MakeTraceID – A unique identifier that connects all segments and subsegments
+// originating from a single client request.
+//  * A trace_id consists of three numbers separated by hyphens. For example,
+//    1-58406520-a006649127e371903a2de979. This includes:
+//  * The version number, that is, 1.
+//  * The time of the original request, in Unix epoch time, in 8 hexadecimal digits.
+//  * For example, 10:00AM December 2nd, 2016 PST in epoch time is 1480615200 seconds,
+//    or 58406520 in hexadecimal.
+//  * A 96-bit identifier for the trace, globally unique, in 24 hexadecimal digits.
+func MakeTraceID(traceID trace.TraceID, startTime time.Time) string {
+	const (
+		digits24 = "123456789012345678901234" // 24 digits
+	)
+
+	var (
+		content = make([]byte, 0, traceIDLength)
+		epoch   = startTime.Unix()
+	)
+
+	content = append(content, "1-"...)
+	content = strconv.AppendInt(content, epoch, 16)
+	content = append(content, "-"...)
+
+	content = append(content, digits24...)                // expand capacity
+	hex.Encode(content[identifierOffset:], traceID[4:16]) // overwrite with identifier
+
+	return string(content)
+}
+
+// ParseAmazonTraceID parses an amazon traceID string in the format 1-5759e988-bd862e3fe1be46a994272793
+func ParseAmazonTraceID(v string) (trace.TraceID, bool) {
+	if len(v) != traceIDLength {
+		return trace.TraceID{}, false
+	}
+
+	epoch, err := hex.DecodeString(v[epochOffset : epochOffset+8])
+	if err != nil {
+		return trace.TraceID{}, false
+	}
+
+	identifier, err := hex.DecodeString(v[identifierOffset:])
+	if err != nil {
+		return trace.TraceID{}, false
+	}
+
+	var traceID trace.TraceID
+	binary.BigEndian.PutUint32(traceID[0:4], binary.BigEndian.Uint32(epoch))
+	for index, b := range identifier {
+		traceID[index+4] = b
+	}
+
+	return traceID, true
+}
+
+// ParseAmazonSpanID parses an amazon spanID
+func ParseAmazonSpanID(v string) (trace.SpanID, bool) {
+	if len(v) != spanIDLength {
+		return trace.SpanID{}, false
+	}
+
+	data, err := hex.DecodeString(v)
+	if err != nil {
+		return trace.SpanID{}, false
+	}
+
+	var spanID trace.SpanID
+	for index, b := range data {
```

Comment:
use `copy(spanID[:], data)` instead of loop

Reply:
thanks.  would you believe I've never used that particular syntax before?

Reply:
dealing with SpanID/TraceID was probably the first time I'd use it too